File: lib/features/home_widgets/domain/models/home_widget_snapshot.dart
import 'dart:convert';

/// Slot identifiers, mirroring the six tiles of the diary's "Heute im Blick"
/// grid. Order is the grid's reading order per column:
/// left column = calories, water, extra — right column = protein, carbs, fat.
class HomeWidgetSlot {
  static const String calories = 'calories';
  static const String water = 'water';
  static const String extra = 'extra';
  static const String protein = 'protein';
  static const String carbs = 'carbs';
  static const String fat = 'fat';

  const HomeWidgetSlot._();
}

/// One progress bar in the widget grid.
///
/// [value] and [target] are already unit-converted for display, and [label] and
/// [unit] are already localized — the Swift side never formats a word.
class HomeWidgetTile {
  final String slot;
  final String label;
  final String unit;
  final double value;
  final double target;

  /// `#RRGGBB`, taken from the very same `Color` the diary bar uses, so the two
  /// can never drift apart.
  final String colorHex;

  const HomeWidgetTile({
    required this.slot,
    required this.label,
    required this.unit,
    required this.value,
    required this.target,
    required this.colorHex,
  });

  Map<String, dynamic> toJson() => {
        'slot': slot,
        'label': label,
        'unit': unit,
        'value': value,
        'target': target,
        'colorHex': colorHex,
      };

  factory HomeWidgetTile.fromJson(Map<String, dynamic> json) => HomeWidgetTile(
        slot: json['slot'] as String,
        label: json['label'] as String,
        unit: json['unit'] as String,
        value: (json['value'] as num).toDouble(),
        target: (json['target'] as num).toDouble(),
        colorHex: json['colorHex'] as String,
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetTile &&
      other.slot == slot &&
      other.label == label &&
      other.unit == unit &&
      other.value == value &&
      other.target == target &&
      other.colorHex == colorHex;

  @override
  int get hashCode => Object.hash(slot, label, unit, value, target, colorHex);
}

/// One of the three readiness pills in the Muscle Readiness widget.
///
/// Mirrors `RecoverySectionCard._buildReadinessPill`: a count, its share of the
/// tracked muscles, and the colour that share is drawn in.
class HomeWidgetRecoveryState {
  /// `recovering` / `ready` / `fresh` — `RecoveryDomainService`'s state keys.
  final String state;

  /// Already localized (`l10n.recoveryStateRecovering` etc.).
  final String label;
  final int count;

  /// Rounded percentage of the tracked muscles, exactly as the card computes it.
  final int percent;
  final String colorHex;

  const HomeWidgetRecoveryState({
    required this.state,
    required this.label,
    required this.count,
    required this.percent,
    required this.colorHex,
  });

  Map<String, dynamic> toJson() => {
        'state': state,
        'label': label,
        'count': count,
        'percent': percent,
        'colorHex': colorHex,
      };

  factory HomeWidgetRecoveryState.fromJson(Map<String, dynamic> json) =>
      HomeWidgetRecoveryState(
        state: json['state'] as String,
        label: json['label'] as String,
        count: (json['count'] as num).toInt(),
        percent: (json['percent'] as num).toInt(),
        colorHex: json['colorHex'] as String,
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetRecoveryState &&
      other.state == state &&
      other.label == label &&
      other.count == count &&
      other.percent == percent &&
      other.colorHex == colorHex;

  @override
  int get hashCode => Object.hash(state, label, count, percent, colorHex);
}

/// The Muscle Readiness widget's payload — a flattened `RecoveryAnalyticsPayload`.
class HomeWidgetRecovery {
  final bool hasData;

  /// Already localized headline (`recoveryOverallLabel`).
  final String headline;

  /// `recoveryOverallColor` as `#RRGGBB`, or null when the app would fall back
  /// to `colorScheme.outline` — the widget resolves that against its own scheme.
  final String? headlineColorHex;
  final List<HomeWidgetRecoveryState> states;

  const HomeWidgetRecovery({
    required this.hasData,
    required this.headline,
    required this.headlineColorHex,
    required this.states,
  });

  Map<String, dynamic> toJson() => {
        'hasData': hasData,
        'headline': headline,
        if (headlineColorHex != null) 'headlineColorHex': headlineColorHex,
        'states': states.map((s) => s.toJson()).toList(),
      };

  factory HomeWidgetRecovery.fromJson(Map<String, dynamic> json) =>
      HomeWidgetRecovery(
        hasData: json['hasData'] as bool,
        headline: json['headline'] as String,
        headlineColorHex: json['headlineColorHex'] as String?,
        states: (json['states'] as List<dynamic>)
            .map((e) =>
                HomeWidgetRecoveryState.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetRecovery &&
      other.hasData == hasData &&
      other.headline == headline &&
      other.headlineColorHex == headlineColorHex &&
      _listEquals(other.states, states);

  @override
  int get hashCode =>
      Object.hash(hasData, headline, headlineColorHex, Object.hashAll(states));
}

/// One bar of the seven-day steps chart.
class HomeWidgetStepsDay {
  /// `yyyy-MM-dd`, so the widget can tell which bar is today without a timezone
  /// of its own.
  final String dayKey;
  final int steps;

  const HomeWidgetStepsDay({required this.dayKey, required this.steps});

  Map<String, dynamic> toJson() => {'dayKey': dayKey, 'steps': steps};

  factory HomeWidgetStepsDay.fromJson(Map<String, dynamic> json) =>
      HomeWidgetStepsDay(
        dayKey: json['dayKey'] as String,
        steps: (json['steps'] as num).toInt(),
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetStepsDay &&
      other.dayKey == dayKey &&
      other.steps == steps;

  @override
  int get hashCode => Object.hash(dayKey, steps);
}

/// The Steps widget's payload, mirroring `StatisticsStepsCard` in its
/// seven-day configuration.
class HomeWidgetSteps {
  /// False when step tracking is off or HealthKit access was never granted —
  /// the widget then shows the permission state rather than a chart of zeros.
  final bool isTrackingEnabled;
  final int todaySteps;
  final int dailyGoal;

  /// Oldest first, always seven entries with the last one being today.
  final List<HomeWidgetStepsDay> days;

  const HomeWidgetSteps({
    required this.isTrackingEnabled,
    required this.todaySteps,
    required this.dailyGoal,
    required this.days,
  });

  Map<String, dynamic> toJson() => {
        'isTrackingEnabled': isTrackingEnabled,
        'todaySteps': todaySteps,
        'dailyGoal': dailyGoal,
        'days': days.map((d) => d.toJson()).toList(),
      };

  factory HomeWidgetSteps.fromJson(Map<String, dynamic> json) =>
      HomeWidgetSteps(
        isTrackingEnabled: json['isTrackingEnabled'] as bool,
        todaySteps: (json['todaySteps'] as num).toInt(),
        dailyGoal: (json['dailyGoal'] as num).toInt(),
        days: (json['days'] as List<dynamic>)
            .map((e) => HomeWidgetStepsDay.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetSteps &&
      other.isTrackingEnabled == isTrackingEnabled &&
      other.todaySteps == todaySteps &&
      other.dailyGoal == dailyGoal &&
      _listEquals(other.days, days);

  @override
  int get hashCode => Object.hash(
        isTrackingEnabled,
        todaySteps,
        dailyGoal,
        Object.hashAll(days),
      );
}

/// One point of a measurement series.
class HomeWidgetMeasurementPoint {
  final double epochMs;

  /// Already converted into the user's display unit.
  final double value;

  const HomeWidgetMeasurementPoint(
      {required this.epochMs, required this.value});

  Map<String, dynamic> toJson() => {'epochMs': epochMs, 'value': value};

  factory HomeWidgetMeasurementPoint.fromJson(Map<String, dynamic> json) =>
      HomeWidgetMeasurementPoint(
        epochMs: (json['epochMs'] as num).toDouble(),
        value: (json['value'] as num).toDouble(),
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetMeasurementPoint &&
      other.epochMs == epochMs &&
      other.value == value;

  @override
  int get hashCode => Object.hash(epochMs, value);
}

/// One selectable metric of the configurable Measurements widget.
///
/// The full series travels rather than one pre-sliced timeframe: the timeframe
/// is a widget configuration the app never learns about, so the widget has to
/// be able to answer for any of the five periods on its own.
class HomeWidgetMeasurementMetric {
  /// The `Measurement.type` key — `weight`, `fat_percent`, `waist`, …
  final String id;

  /// Already localized (`l10n.getLocalizedMeasurementName`).
  final String name;

  /// Display unit suffix (`kg`, `%`, `cm`), matching `_getMeasurementUnit`.
  final String unit;

  /// Oldest first.
  final List<HomeWidgetMeasurementPoint> points;

  const HomeWidgetMeasurementMetric({
    required this.id,
    required this.name,
    required this.unit,
    required this.points,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'unit': unit,
        'points': points.map((p) => p.toJson()).toList(),
      };

  factory HomeWidgetMeasurementMetric.fromJson(Map<String, dynamic> json) =>
      HomeWidgetMeasurementMetric(
        id: json['id'] as String,
        name: json['name'] as String,
        unit: json['unit'] as String,
        points: (json['points'] as List<dynamic>)
            .map((e) =>
                HomeWidgetMeasurementPoint.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetMeasurementMetric &&
      other.id == id &&
      other.name == name &&
      other.unit == unit &&
      _listEquals(other.points, points);

  @override
  int get hashCode => Object.hash(id, name, unit, Object.hashAll(points));
}

/// The Last Workout widget's payload.
class HomeWidgetLastWorkout {
  final int id;
  final String title;
  final double completedAtEpochMs;
  final int durationSeconds;

  /// Already converted into the user's weight unit. Null for a session without
  /// a single weighted set — the widget shows reps instead.
  final double? totalVolume;
  final String volumeUnit;
  final int totalReps;
  final int totalSets;

  /// File name inside the App Group container of the front+back heatmap the app
  /// renders when a workout is finished, or null if it could not be produced.
  final String? heatmapImageName;

  const HomeWidgetLastWorkout({
    required this.id,
    required this.title,
    required this.completedAtEpochMs,
    required this.durationSeconds,
    required this.totalVolume,
    required this.volumeUnit,
    required this.totalReps,
    required this.totalSets,
    required this.heatmapImageName,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'completedAtEpochMs': completedAtEpochMs,
        'durationSeconds': durationSeconds,
        if (totalVolume != null) 'totalVolume': totalVolume,
        'volumeUnit': volumeUnit,
        'totalReps': totalReps,
        'totalSets': totalSets,
        if (heatmapImageName != null) 'heatmapImageName': heatmapImageName,
      };

  factory HomeWidgetLastWorkout.fromJson(Map<String, dynamic> json) =>
      HomeWidgetLastWorkout(
        id: (json['id'] as num).toInt(),
        title: json['title'] as String,
        completedAtEpochMs: (json['completedAtEpochMs'] as num).toDouble(),
        durationSeconds: (json['durationSeconds'] as num).toInt(),
        totalVolume: (json['totalVolume'] as num?)?.toDouble(),
        volumeUnit: json['volumeUnit'] as String,
        totalReps: (json['totalReps'] as num).toInt(),
        totalSets: (json['totalSets'] as num).toInt(),
        heatmapImageName: json['heatmapImageName'] as String?,
      );

  @override
  bool operator ==(Object other) =>
      other is HomeWidgetLastWorkout &&
      other.id == id &&
      other.title == title &&
      other.completedAtEpochMs == completedAtEpochMs &&
      other.durationSeconds == durationSeconds &&
      other.totalVolume == totalVolume &&
      other.volumeUnit == volumeUnit &&
      other.totalReps == totalReps &&
      other.totalSets == totalSets &&
      other.heatmapImageName == heatmapImageName;

  @override
  int get hashCode => Object.hash(
        id,
        title,
        completedAtEpochMs,
        durationSeconds,
        totalVolume,
        volumeUnit,
        totalReps,
        totalSets,
        heatmapImageName,
      );
}

bool _listEquals<T>(List<T> a, List<T> b) {
  if (identical(a, b)) return true;
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

/// The payload published to the iOS App Group.
///
/// Aggregate totals and targets only — no food names, no timestamps, no entry
/// level data leaves the app's own container.
class HomeWidgetSnapshot {
  /// v2 added the four statistics sections below. Every one of them is nullable
  /// and every reader treats a missing section as "this widget has nothing to
  /// show yet", so a v1 payload left over from an older build still decodes.
  static const int currentSchemaVersion = 2;

  /// The hour at which the diary rolls over to the next day. Mirrors
  /// `resolveDiaryInitialDate`, and travels in the payload so the widget never
  /// hardcodes its own copy of the rule.
  static const int diaryRolloverHour = 3;

  final int schemaVersion;
  final double generatedAtEpochMs;

  /// `yyyy-MM-dd` in the user's local calendar.
  final String logicalDayKey;
  final int rolloverHour;
  final bool isAiEnabled;
  final List<HomeWidgetTile> tiles;

  /// Null until the app has computed recovery analytics at least once.
  final HomeWidgetRecovery? recovery;
  final HomeWidgetSteps? steps;

  /// Every metric the user has ever recorded, so the configurable widget can
  /// offer them all without a round trip into the app.
  final List<HomeWidgetMeasurementMetric> measurements;
  final HomeWidgetLastWorkout? lastWorkout;

  const HomeWidgetSnapshot({
    required this.schemaVersion,
    required this.generatedAtEpochMs,
    required this.logicalDayKey,
    required this.rolloverHour,
    required this.isAiEnabled,
    required this.tiles,
    this.recovery,
    this.steps,
    this.measurements = const [],
    this.lastWorkout,
  });

  Map<String, dynamic> toJson() => {
        'schemaVersion': schemaVersion,
        'generatedAtEpochMs': generatedAtEpochMs,
        'logicalDayKey': logicalDayKey,
        'rolloverHour': rolloverHour,
        'isAiEnabled': isAiEnabled,
        'tiles': tiles.map((t) => t.toJson()).toList(),
        if (recovery != null) 'recovery': recovery!.toJson(),
        if (steps != null) 'steps': steps!.toJson(),
        if (measurements.isNotEmpty)
          'measurements': measurements.map((m) => m.toJson()).toList(),
        if (lastWorkout != null) 'lastWorkout': lastWorkout!.toJson(),
      };

  factory HomeWidgetSnapshot.fromJson(Map<String, dynamic> json) =>
      HomeWidgetSnapshot(
        schemaVersion: json['schemaVersion'] as int,
        generatedAtEpochMs: (json['generatedAtEpochMs'] as num).toDouble(),
        logicalDayKey: json['logicalDayKey'] as String,
        rolloverHour: json['rolloverHour'] as int,
        isAiEnabled: json['isAiEnabled'] as bool,
        tiles: (json['tiles'] as List<dynamic>)
            .map((e) => HomeWidgetTile.fromJson(e as Map<String, dynamic>))
            .toList(),
        recovery: json['recovery'] == null
            ? null
            : HomeWidgetRecovery.fromJson(
                json['recovery'] as Map<String, dynamic>),
        steps: json['steps'] == null
            ? null
            : HomeWidgetSteps.fromJson(json['steps'] as Map<String, dynamic>),
        measurements: (json['measurements'] as List<dynamic>? ?? const [])
            .map((e) =>
                HomeWidgetMeasurementMetric.fromJson(e as Map<String, dynamic>))
            .toList(),
        lastWorkout: json['lastWorkout'] == null
            ? null
            : HomeWidgetLastWorkout.fromJson(
                json['lastWorkout'] as Map<String, dynamic>),
      );

  String encode() => jsonEncode(toJson());

  static HomeWidgetSnapshot decode(String source) =>
      HomeWidgetSnapshot.fromJson(jsonDecode(source) as Map<String, dynamic>);
}

File: android/app/src/main/kotlin/com/rfivesix/trainlibre/widgets/snapshot/HomeWidgetSnapshot.kt
package com.rfivesix.trainlibre.widgets.snapshot

import java.math.BigDecimal
import java.math.RoundingMode
import java.util.Calendar
import java.util.Locale
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Shared decoder for the snapshot the Flutter app publishes.
 *
 * `ignoreUnknownKeys` is what lets an older build survive a newer app: the Dart
 * side may add a section before the widget knows about it, and a widget that
 * refused to decode would freeze on its last good render instead.
 */
internal val homeWidgetJson: Json = Json {
    ignoreUnknownKeys = true
}

/**
 * Slot identifiers of the diary's "Heute im Blick" grid, in its reading order
 * per column: left = calories, water, extra — right = protein, carbs, fat.
 */
object HomeWidgetSlot {
    const val CALORIES = "calories"
    const val WATER = "water"
    const val EXTRA = "extra"
    const val PROTEIN = "protein"
    const val CARBS = "carbs"
    const val FAT = "fat"

    /** The grid's reading order, used when no configuration has been made yet. */
    val defaultOrder = listOf(CALORIES, PROTEIN, WATER, CARBS, EXTRA, FAT)
}

/**
 * One progress bar in the widget grid.
 *
 * Everything localized or unit-dependent (`label`, `unit`, the already converted
 * `value`) is computed in Dart. The widget only formats numbers.
 */
@Serializable
data class HomeWidgetTile(
    val slot: String,
    val label: String,
    val unit: String,
    val value: Double,
    val target: Double,
    /** `#RRGGBB`, taken from the very same colour the diary bar uses. */
    val colorHex: String,
) {
    val hasTarget: Boolean get() = target > 0

    /** Clamped 0..1, mirroring `GlassProgressBar`'s `rawProgress.clamp(0, 1)`. */
    val progress: Float
        get() = if (!hasTarget) 0f else (value / target).coerceIn(0.0, 1.0).toFloat()

    /**
     * Byte-identical to the Dart side:
     * `'${value.toStringAsFixed(1)} / ${target.toStringAsFixed(0)} $unit'`.
     */
    val valueText: String
        get() {
            val left = dartFixed(value, 1)
            if (!hasTarget) return "$left $unit"
            return "$left / ${dartFixed(target, 0)} $unit"
        }

    fun zeroed(): HomeWidgetTile = copy(value = 0.0)

    companion object {
        /**
         * `Double.toStringAsFixed` from Dart, reproduced exactly.
         *
         * `String.format("%.1f")` is not equivalent: Java formats from the
         * double's *shortest* decimal representation, so it sees 0.15 and rounds
         * up, while Dart rounds the double's true value (0.1499…) down.
         * `BigDecimal(double)` is that true value, and HALF_UP rounds ties away
         * from zero the way Dart does — so 40.25 kg of volume renders as 40.3 in
         * both the app and the widget.
         */
        fun dartFixed(value: Double, digits: Int): String {
            if (value.isNaN() || value.isInfinite()) {
                return String.format(Locale.US, "%.${digits}f", value)
            }
            // BigDecimal has no negative zero, so "-0.0" cannot appear here any
            // more than it does in Dart.
            return BigDecimal(value).setScale(digits, RoundingMode.HALF_UP).toPlainString()
        }
    }
}

// --- Statistics sections (schema v2) ---

/**
 * One readiness pill of the Muscle Readiness widget.
 *
 * Mirrors `RecoverySectionCard._buildReadinessPill`. The count, its share and
 * the colour all arrive precomputed — the widget only lays them out.
 */
@Serializable
data class HomeWidgetRecoveryState(
    /** `recovering` / `ready` / `fresh` — `RecoveryDomainService`'s state keys. */
    val state: String,
    val label: String,
    val count: Int,
    val percent: Int,
    val colorHex: String,
)

@Serializable
data class HomeWidgetRecovery(
    val hasData: Boolean,
    val headline: String,
    /**
     * Absent when the app would have used `colorScheme.outline` — the widget has
     * its own scheme and resolves that itself rather than being handed a colour
     * from the app's theme.
     */
    val headlineColorHex: String? = null,
    val states: List<HomeWidgetRecoveryState> = emptyList(),
)

@Serializable
data class HomeWidgetStepsDay(
    /**
     * `yyyy-MM-dd` in the user's calendar, so "which bar is today" needs no
     * timezone arithmetic in the widget.
     */
    val dayKey: String,
    val steps: Int,
)

@Serializable
data class HomeWidgetSteps(
    /**
     * False when step tracking is off or permission was never granted — the
     * widget then shows that state rather than a chart of zeros.
     */
    val isTrackingEnabled: Boolean,
    val todaySteps: Int,
    val dailyGoal: Int,
    /** Oldest first, seven entries, the last one being today. */
    val days: List<HomeWidgetStepsDay> = emptyList(),
) {
    /**
     * `StatisticsStepsCard`'s scale: the taller of the goal and the best day, so
     * no bar can leave the chart no matter how big a day was.
     */
    val chartMaximum: Int
        get() = maxOf(days.maxOfOrNull { it.steps } ?: 0, maxOf(dailyGoal, 1))
}

@Serializable
data class HomeWidgetMeasurementPoint(
    val epochMs: Double,
    /** Already converted into the user's display unit. */
    val value: Double,
)

/**
 * One selectable metric of the configurable Measurements widget.
 *
 * Carries the whole series rather than a pre-sliced timeframe: the timeframe is a
 * widget configuration the app never sees, so the widget has to be able to answer
 * for any of the five periods on its own.
 */
@Serializable
data class HomeWidgetMeasurementMetric(
    /** The `Measurement.type` key — `weight`, `fat_percent`, `waist`, … */
    val id: String,
    /** Already localized (`l10n.getLocalizedMeasurementName`). */
    val name: String,
    /** Display unit suffix (`kg`, `%`, `cm`). */
    val unit: String,
    /** Oldest first. */
    val points: List<HomeWidgetMeasurementPoint> = emptyList(),
) {
    /**
     * The points inside the last [days], or all of them when [days] is null (the
     * "Max" timeframe).
     */
    fun pointsWithinDays(
        days: Int?,
        nowMs: Long = System.currentTimeMillis(),
    ): List<HomeWidgetMeasurementPoint> {
        if (days == null) return points
        val cutoff = nowMs - days.toLong() * 86_400_000L
        return points.filter { it.epochMs >= cutoff }
    }
}

@Serializable
data class HomeWidgetLastWorkout(
    val id: Int,
    val title: String,
    val completedAtEpochMs: Double,
    val durationSeconds: Int,
    /**
     * Absent for a session without a single weighted set — the widget shows the
     * rep count in that slot instead.
     */
    val totalVolume: Double? = null,
    val volumeUnit: String,
    val totalReps: Int,
    val totalSets: Int,
    /**
     * File name inside the widget container, written by the app when the workout
     * was finished.
     */
    val heatmapImageName: String? = null,
)

/**
 * The full payload the app publishes.
 *
 * Holds aggregate totals and targets only — no food names, no timestamps, no
 * entry level data ever reaches the widget container.
 */
@Serializable
data class HomeWidgetSnapshot(
    val schemaVersion: Int,
    val generatedAtEpochMs: Double,
    /**
     * The diary day these totals belong to, as `yyyy-MM-dd` in the user's local
     * calendar.
     */
    val logicalDayKey: String,
    /**
     * The hour at which the app rolls the diary over to the next day (3).
     * Transmitted rather than hardcoded so the widget cannot drift from
     * `resolveDiaryInitialDate`.
     */
    val rolloverHour: Int,
    val isAiEnabled: Boolean,
    val tiles: List<HomeWidgetTile> = emptyList(),
    // Every section below is optional, and every reader treats a missing one as
    // "nothing to show yet" — that is what lets a v1 payload left over from an
    // older build still decode.
    val recovery: HomeWidgetRecovery? = null,
    val steps: HomeWidgetSteps? = null,
    val measurements: List<HomeWidgetMeasurementMetric> = emptyList(),
    val lastWorkout: HomeWidgetLastWorkout? = null,
) {
    /**
     * The same totals reset to zero but the targets kept, for a day the app has
     * not written a snapshot for yet. This is what makes the 03:00 rollover work
     * while the app is closed: nothing can have been logged for the new day
     * without the app running, so zero is not a guess — it is the answer.
     *
     * Only the diary rolls over. Recovery, steps, measurements and the last
     * workout are not "today's totals" — they carry across the boundary
     * unchanged, and blanking them would be a lie, not a reset.
     */
    fun zeroed(forDayKey: String): HomeWidgetSnapshot =
        copy(logicalDayKey = forDayKey, tiles = tiles.map { it.zeroed() })

    /**
     * The snapshot as the widget should show it *now*.
     *
     * A widget can outlive the app's last write by hours, so if the rollover has
     * passed since then, the stale diary totals are the wrong answer.
     */
    fun resolvedFor(nowMs: Long = System.currentTimeMillis()): HomeWidgetSnapshot {
        val today = HomeWidgetDay.dayKey(nowMs, rolloverHour)
        return if (today == logicalDayKey) this else zeroed(today)
    }

    companion object {
        /** v2 added the four statistics sections. */
        const val CURRENT_SCHEMA_VERSION = 2

        /** The hour at which the diary rolls over. Mirrors `resolveDiaryInitialDate`. */
        const val DIARY_ROLLOVER_HOUR = 3

        /**
         * Returns null rather than throwing: every caller is a widget render that
         * has to draw *something*, and an empty state beats a crash in the
         * launcher's host process.
         */
        fun decodeOrNull(source: String): HomeWidgetSnapshot? = try {
            homeWidgetJson.decodeFromString<HomeWidgetSnapshot>(source)
        } catch (_: Exception) {
            null
        }
    }
}

/**
 * Resolves which diary day a point in time belongs to.
 *
 * The rule lives in Dart (`resolveDiaryInitialDate`): before 03:00 the diary
 * still shows the previous day. `rolloverHour` carries it across, so this stays a
 * single rule with one owner.
 */
object HomeWidgetDay {
    fun dayKey(atMs: Long, rolloverHour: Int): String {
        val calendar = Calendar.getInstance().apply {
            timeInMillis = atMs
            add(Calendar.HOUR_OF_DAY, -rolloverHour)
        }
        return String.format(
            Locale.US,
            "%04d-%02d-%02d",
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH) + 1,
            calendar.get(Calendar.DAY_OF_MONTH),
        )
    }

    /**
     * The next instant at which [dayKey] changes its answer — today's boundary if
     * it is still ahead of us, tomorrow's otherwise.
     */
    fun nextRollover(afterMs: Long, rolloverHour: Int): Long {
        val boundary = Calendar.getInstance().apply {
            timeInMillis = afterMs
            set(Calendar.HOUR_OF_DAY, rolloverHour % 24)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        if (boundary.timeInMillis > afterMs) return boundary.timeInMillis
        boundary.add(Calendar.DAY_OF_MONTH, 1)
        return boundary.timeInMillis
    }
}
